--- app/src/main/cpp/third_party/quickjs/quickjs.c
+++ app/src/main/cpp/third_party/quickjs/quickjs.c
@@ -1537,6 +1537,8 @@
                                        JSValueConst this_val,
                                        int argc, JSValueConst *argv, int flags);
 static void js_c_closure_finalizer(JSRuntime *rt, JSValueConst val);
+static void js_c_closure_mark(JSRuntime *rt, JSValueConst val,
+                              JS_MarkFunc *mark_func);
 static JSValue js_call_c_closure(JSContext *ctx, JSValueConst func_obj,
                                  JSValueConst this_val,
                                  int argc, JSValueConst *argv, int flags);
@@ -2231,7 +2233,7 @@
     { JS_ATOM_Function, js_bytecode_function_finalizer, js_bytecode_function_mark }, /* JS_CLASS_BYTECODE_FUNCTION */
     { JS_ATOM_Function, js_bound_function_finalizer, js_bound_function_mark }, /* JS_CLASS_BOUND_FUNCTION */
     { JS_ATOM_Function, js_c_function_data_finalizer, js_c_function_data_mark }, /* JS_CLASS_C_FUNCTION_DATA */
-    { JS_ATOM_Function, js_c_closure_finalizer, NULL},                           /* JS_CLASS_C_CLOSURE */
+    { JS_ATOM_Function, js_c_closure_finalizer, js_c_closure_mark},                           /* JS_CLASS_C_CLOSURE */
     { JS_ATOM_GeneratorFunction, js_bytecode_function_finalizer, js_bytecode_function_mark },  /* JS_CLASS_GENERATOR_FUNCTION */
     { JS_ATOM_ForInIterator, js_for_in_iterator_finalizer, js_for_in_iterator_mark },      /* JS_CLASS_FOR_IN_ITERATOR */
     { JS_ATOM_RegExp, js_regexp_finalizer, NULL },                              /* JS_CLASS_REGEXP */
@@ -6551,6 +6553,7 @@
 }

 typedef struct JSCFunctionDataRecord {
+    JSContext *realm;
     JSCFunctionData *func;
     uint8_t length;
     uint8_t data_len;
@@ -6563,6 +6566,7 @@
     int i;

     if (s) {
+        JS_FreeContext(s->realm);
         for(i = 0; i < s->data_len; i++) {
             JS_FreeValueRT(rt, s->data[i]);
         }
@@ -6577,6 +6581,7 @@
     int i;

     if (s) {
+        mark_func(rt, &s->realm->header);
         for(i = 0; i < s->data_len; i++) {
             JS_MarkValue(rt, s->data[i], mark_func);
         }
@@ -6615,7 +6620,7 @@
     prev_sf = rt->current_stack_frame;
     sf->prev_frame = prev_sf;
     rt->current_stack_frame = sf;
-    // TODO(bnoordhuis) switch realms like js_call_c_function does
+    ctx = s->realm; /* change the current realm */
     sf->is_strict_mode = false;
     sf->is_constructor = (flags & JS_CALL_FLAG_CONSTRUCTOR) != 0;
     sf->cur_func = unsafe_unconst(func_obj);
@@ -6644,7 +6649,8 @@
         JS_FreeValue(ctx, func_obj);
         return JS_EXCEPTION;
     }
-    s->func = func;
+    s->realm = JS_DupContext(ctx);
+    s->func = func;
     s->length = length;
     s->data_len = data_len;
     s->magic = magic;
@@ -6693,6 +6699,7 @@
 }

 typedef struct JSCClosureRecord {
+    JSContext *realm;
     JSCClosure *func;
     uint16_t length;
     uint16_t magic;
@@ -6704,6 +6711,7 @@
     JSCClosureRecord *s = JS_GetOpaque(val, JS_CLASS_C_CLOSURE);

     if (s) {
+        JS_FreeContext(s->realm);
         if (s->opaque_finalize)
            s->opaque_finalize(s->opaque);

@@ -6711,6 +6719,15 @@
     }
 }

+static void js_c_closure_mark(JSRuntime *rt, JSValueConst val,
+                              JS_MarkFunc *mark_func)
+{
+    JSCClosureRecord *s = JS_GetOpaque(val, JS_CLASS_C_CLOSURE);
+    if (s) {
+        mark_func(rt, &s->realm->header);
+    }
+}
+
 static JSValue js_call_c_closure(JSContext *ctx, JSValueConst func_obj,
                                  JSValueConst this_val,
                                  int argc, JSValueConst *argv, int flags)
@@ -6741,7 +6758,7 @@
     prev_sf = rt->current_stack_frame;
     sf->prev_frame = prev_sf;
     rt->current_stack_frame = sf;
-    // TODO(bnoordhuis) switch realms like js_call_c_function does
+    ctx = s->realm; /* change the current realm */
     sf->is_strict_mode = false;
     sf->is_constructor = (flags & JS_CALL_FLAG_CONSTRUCTOR) != 0;
     sf->cur_func = unsafe_unconst(func_obj);
@@ -6769,7 +6786,8 @@
         JS_FreeValue(ctx, func_obj);
         return JS_EXCEPTION;
     }
-    s->func = func;
+    s->realm = JS_DupContext(ctx);
+    s->func = func;
     s->length = length;
     s->magic = magic;
     s->opaque = opaque;
@@ -20969,6 +20987,12 @@
     case JS_CLASS_C_FUNCTION:
         realm = p->u.cfunc.realm;
         break;
+    case JS_CLASS_C_FUNCTION_DATA:
+        realm = p->u.c_function_data_record->realm;
+        break;
+    case JS_CLASS_C_CLOSURE:
+        realm = p->u.c_closure_record->realm;
+        break;
     case JS_CLASS_BYTECODE_FUNCTION:
     case JS_CLASS_GENERATOR_FUNCTION:
     case JS_CLASS_ASYNC_FUNCTION:
